pir/cpp/server.cpp

```diff
@@ -47,9 +48,85 @@ StatusOr<std::unique_ptr<PIRServer>> PIRServer::Create(
 
 StatusOr<PIRPayload> PIRServer::ProcessRequest(
     const PIRPayload& payload) const {
-  ASSIGN_OR_RETURN(auto result, db_->multiply(payload.Get()));
+  if (payload.Get().size() != 1) {
+    return InvalidArgumentError("Number of ciphertexts in request must be 1");
+  }
+  if (!payload.GetKeys()) {
+    return InvalidArgumentError("Must have Galois Keys in request");
+  }
+
+  auto selection_vector =
+      oblivious_expansion(payload.Get()[0], DBSize(), *payload.GetKeys());
+
+  ASSIGN_OR_RETURN(auto mult_results, db_->multiply(selection_vector));
+
+  seal::Ciphertext result;
+  context_->Evaluator()->add_many(mult_results, result);
+
+  return PIRPayload::Load(vector<seal::Ciphertext>{result});
+}
+
+void PIRServer::substitute_power_x_inplace(
+    seal::Ciphertext& ct, uint32_t power,
+    const seal::GaloisKeys& gal_keys) const {
+  context_->Evaluator()->apply_galois_inplace(ct, power, gal_keys);
+}
+
+void PIRServer::multiply_power_of_x(const seal::Ciphertext& encrypted, int k,
+                                    seal::Ciphertext& destination) const {
+  // This has to get the actual params from the SEALContext. Using just the
+  // params from PIR doesn't work.
+  const auto& params = context_->SEALContext()->first_context_data()->parms();
+  const auto poly_modulus_degree = params.poly_modulus_degree();
+  const auto coeff_mod_count = params.coeff_modulus().size();
 
-  return PIRPayload::Load(result);
+  // handle negative values of k properly
+  uint32_t index = (k >= 0) ? k : (poly_modulus_degree * 2 + k);
+
+  // have to make a copy here
+  destination = encrypted;
+
+  // Loop over polynomials in ciphertext
+  for (int i = 0; i < encrypted.size(); i++) {
+    // loop over each coefficient in polynomial
+    for (int j = 0; j < coeff_mod_count; j++) {
+      seal::util::negacyclic_shift_poly_coeffmod(
+          encrypted.data(i) + (j * poly_modulus_degree), poly_modulus_degree,
+          index, params.coeff_modulus()[j],
+          destination.data(i) + (j * poly_modulus_degree));
+    }
+  }
+}
+
+std::vector<seal::Ciphertext> PIRServer::oblivious_expansion(
+    const seal::Ciphertext& ct, const size_t num_items,
+    const seal::GaloisKeys gal_keys) const {
+  const auto poly_modulus_degree =
+      context_->Parameters()->GetEncryptionParams().poly_modulus_degree();
+  size_t logm = ceil(log2(num_items));
+  std::vector<seal::Ciphertext> results(1 << logm);
+  results[0] = ct;
+
+  for (size_t j = 0; j < logm; ++j) {
+    const size_t two_power_j = (1 << j);
+    for (size_t k = 0; k < two_power_j; ++k) {
+      auto c0 = results[k];
+      seal::Ciphertext c1;
```

Comment:
Shouldn't this ciphertext be generated from the same SEAL Context?

Reply:
Not sure what you mean. Ciphertext doesn't have any connection to a SEAL Context, does it?

Reply:
Yup. When the copy happens in multiply_power_of_x, everything will be set properly. We don't need to reserve here.